File: Sources/Lumina/Protocol.swift
// Sources/Lumina/Protocol.swift
import Foundation

// MARK: - Host Messages (sent to guest)

public enum HostMessage: Sendable {
    case exec(id: String, cmd: String, timeout: Int, env: [String: String], cwd: String? = nil)
    case upload(path: String, data: String, mode: String, seq: Int, eof: Bool)
    case downloadReq(path: String)
    /// Send a signal to a running command (by id) or all commands (id nil).
    /// The guest sends `signal` to the process group, waits `gracePeriod` seconds, then SIGKILL.
    case cancel(id: String?, signal: Int32, gracePeriod: Int)
    /// Send stdin data to a running command identified by id.
    case stdin(id: String, data: String)
    /// Close the stdin pipe for a running command.
    case stdinClose(id: String)
    /// Configure guest network (host-driven, Apple-style).
    case configureNetwork(ip: String, gateway: String, dns: String)
    /// Start a PTY-backed interactive command. Output flows back as `.ptyOutput`
    /// and the process exits via the shared `.exit(id:code:)` message.
    case ptyExec(id: String, cmd: String, timeout: Int, env: [String: String], cols: Int, rows: Int)
    /// Write base64-encoded bytes to a running PTY's master side (keyboard input).
    case ptyInput(id: String, data: String)
    /// Resize the window of a running PTY (sends TIOCSWINSZ on the guest).
    case windowResize(id: String, cols: Int, rows: Int)
}

// MARK: - Guest Messages (received from guest)

public enum GuestMessage: Sendable, Equatable {
    case ready
    /// Text output (UTF-8). `data` carries the text verbatim.
    case output(id: String, stream: OutputStream, data: String)
    /// Binary output. `bytes` carries the raw bytes that were base64-encoded
    /// on the wire. Produced when the guest detects a non-UTF-8 chunk via
    /// `utf8.Valid()` before emitting.
    case outputBinary(id: String, stream: OutputStream, bytes: Data)
    case exit(id: String, code: Int32)
    case heartbeat
    case uploadAck(seq: Int)
    case uploadDone(path: String)
    case uploadError(path: String, error: String)
    case downloadData(path: String, data: String, seq: Int, eof: Bool)
    case downloadError(path: String, error: String)
    case networkReady(ip: String)
    /// PTY-backed output (merged stdout+stderr). `data` is the base64-encoded raw bytes
    /// read from the PTY master. Exit is delivered via the shared `.exit(id:code:)` case.
    case ptyOutput(id: String, data: String)
}

public enum OutputStream: String, Sendable, Equatable, Codable {
    case stdout
    case stderr
}

// MARK: - Wire Format

enum LuminaProtocol {
    // 128KB — must accommodate 48KB raw chunks from guest agent (48*4/3 ≈ 64KB base64 + JSON envelope).
    static let maxMessageSize = 131_072

    static func encode(_ message: HostMessage) throws -> Data {
        let dict: [String: Any]
        switch message {
        case .exec(let id, let cmd, let timeout, let env, let cwd):
            var d: [String: Any] = ["type": "exec", "id": id, "cmd": cmd, "timeout": timeout, "env": env]
            if let cwd = cwd { d["cwd"] = cwd }
            dict = d
        case .upload(let path, let dataStr, let mode, let seq, let eof):
            dict = ["type": "upload", "path": path, "data": dataStr, "mode": mode, "seq": seq, "eof": eof]
        case .downloadReq(let path):
            dict = ["type": "download_req", "path": path]
        case .cancel(let id, let signal, let gracePeriod):
            var d: [String: Any] = ["type": "cancel", "signal": Int(signal), "grace_period": gracePeriod]
            if let id = id { d["id"] = id }
            dict = d
        case .stdin(let id, let data):
            dict = ["type": "stdin", "id": id, "data": data]
        case .stdinClose(let id):
            dict = ["type": "stdin_close", "id": id]
        case .configureNetwork(let ip, let gateway, let dns):
            dict = ["type": "configure_network", "ip": ip, "gateway": gateway, "dns": dns]
        case .ptyExec(let id, let cmd, let timeout, let env, let cols, let rows):
            dict = ["type": "pty_exec", "id": id, "cmd": cmd, "timeout": timeout, "env": env, "cols": cols, "rows": rows]
        case .ptyInput(let id, let data):
            dict = ["type": "pty_input", "id": id, "data": data]
        case .windowResize(let id, let cols, let rows):
            dict = ["type": "window_resize", "id": id, "cols": cols, "rows": rows]
        }
        var data = try JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys])
        data.append(contentsOf: [UInt8(ascii: "\n")])
        return data
    }

    static func decodeGuest(_ data: Data) throws -> GuestMessage {
        let trimmed = data.prefix(while: { $0 != UInt8(ascii: "\n") })
        guard let json = try? JSONSerialization.jsonObject(with: trimmed) as? [String: Any],
              let type = json["type"] as? String
        else {
            throw LuminaError.protocolError("Invalid JSON in guest message")
        }

        switch type {
        case "ready":
            return .ready
        case "output":
            guard let id = json["id"] as? String,
                  let streamStr = json["stream"] as? String,
                  let stream = OutputStream(rawValue: streamStr),
                  let outputData = json["data"] as? String
            else {
                throw LuminaError.protocolError("Malformed output message")
            }
            // Backward-compatible binary-output envelope. Old guests never set
            // "encoding" → treated as UTF-8 text. New guests emitting non-UTF-8
            // chunks set "encoding":"base64" and base64-encode the bytes in the
            // "data" string.
            if let encoding = json["encoding"] as? String, encoding == "base64" {
                guard let bytes = Data(base64Encoded: outputData) else {
                    throw LuminaError.protocolError("Malformed base64 output data")
                }
                return .outputBinary(id: id, stream: stream, bytes: bytes)
            }
            return .output(id: id, stream: stream, data: outputData)
        case "exit":
            guard let id = json["id"] as? String,
                  let code = json["code"] as? Int else {
                throw LuminaError.protocolError("Malformed exit message: missing id or code")
            }
            return .exit(id: id, code: Int32(code))
        case "heartbeat":
            return .heartbeat
        case "upload_ack":
            guard let seq = json["seq"] as? Int else {
                throw LuminaError.protocolError("Malformed upload_ack: missing seq")
            }
            return .uploadAck(seq: seq)
        case "upload_done":
            guard let path = json["path"] as? String else {
                throw LuminaError.protocolError("Malformed upload_done: missing path")
            }
            return .uploadDone(path: path)
        case "upload_error":
            guard let path = json["path"] as? String,
                  let errorStr = json["error"] as? String else {
                throw LuminaError.protocolError("Malformed upload_error: missing path/error")
            }
            return .uploadError(path: path, error: errorStr)
        case "download_data":
            guard let path = json["path"] as? String,
                  let dataStr = json["data"] as? String,
                  let seq = json["seq"] as? Int,
                  let eof = json["eof"] as? Bool else {
                throw LuminaError.protocolError("Malformed download_data: missing fields")
            }
            return .downloadData(path: path, data: dataStr, seq: seq, eof: eof)
        case "download_error":
            guard let path = json["path"] as? String,
                  let errorStr = json["error"] as? String else {
                throw LuminaError.protocolError("Malformed download_error: missing path/error")
            }
            return .downloadError(path: path, error: errorStr)
        case "network_ready":
            let ip = json["ip"] as? String ?? ""
            return .networkReady(ip: ip)
        case "pty_output":
            guard let id = json["id"] as? String,
                  let outputData = json["data"] as? String else {
                throw LuminaError.protocolError("Malformed pty_output: missing id or data")
            }
            return .ptyOutput(id: id, data: outputData)
        default:
            throw LuminaError.protocolError("Unknown message type: \(type)")
        }
    }
}

// MARK: - Public alias using backtick-escaped name

/// Namespace alias to allow `Protocol.encode` / `Protocol.decodeGuest` call sites.
typealias Protocol = LuminaProtocol

File: Guest/lumina-agent/internal/protocol/protocol.go
// Package protocol defines the NDJSON wire types exchanged between the
// host (Swift library) and the guest agent. Every message carries a
// `type` discriminator; the host dispatches on it.
//
// Keep this file purely declarative — no behaviour, no I/O. It is
// imported by every other internal package and by the host-side test
// suite (via the fixture golden bytes).
package protocol

// ── Wire-level constants ────────────────────────────────────────────

const (
	// MaxChunkSize is the ceiling on a single NDJSON frame. Matches
	// the host-side SessionServer limit and the Swift ProtocolTests
	// expectations.
	MaxChunkSize = 65536 // 64 KiB
)

// Type discriminators. Keep in one place so typos are a compile error,
// not a runtime dispatch miss.
const (
	TypeReady            = "ready"
	TypeHeartbeat        = "heartbeat"
	TypeExec             = "exec"
	TypeOutput           = "output"
	TypeExit             = "exit"
	TypeStdin            = "stdin"
	TypeStdinClose       = "stdin_close"
	TypeCancel           = "cancel"
	TypeConfigureNetwork = "configure_network"
	TypeNetworkReady     = "network_ready"
	TypeNetworkError     = "network_error"
	TypeNetworkMetrics   = "network_metrics"

	TypeUpload        = "upload"
	TypeUploadAck     = "upload_ack"
	TypeUploadError   = "upload_error"
	TypeUploadDone    = "upload_done"
	TypeDownloadReq   = "download_req"
	TypeDownloadData  = "download_data"
	TypeDownloadError = "download_error"

	TypePtyExec      = "pty_exec"
	TypePtyInput     = "pty_input"
	TypePtyOutput    = "pty_output"
	TypeWindowResize = "window_resize"

	TypePortForwardStart = "port_forward_start"
	TypePortForwardStop  = "port_forward_stop"
	TypePortForwardReady = "port_forward_ready"
	TypePortForwardError = "port_forward_error"
)

// Stream names used in OutputMsg.
const (
	StreamStdout = "stdout"
	StreamStderr = "stderr"
)

// Envelope is the minimal header used to peek at the Type before
// decoding into a concrete struct.
type Envelope struct {
	Type string `json:"type"`
}

// ── Host → guest messages ──────────────────────────────────────────

type ExecRequest struct {
	Type    string            `json:"type"`
	ID      string            `json:"id"`
	Cmd     string            `json:"cmd"`
	Timeout int               `json:"timeout"`
	Env     map[string]string `json:"env"`
	Cwd     string            `json:"cwd,omitempty"`
}

type StdinMsg struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Data string `json:"data"`
}

type StdinCloseMsg struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

type CancelMsg struct {
	Type        string `json:"type"`
	ID          string `json:"id,omitempty"`
	Signal      int    `json:"signal"`
	GracePeriod int    `json:"grace_period"`
}

type ConfigureNetworkMsg struct {
	Type    string `json:"type"`
	IP      string `json:"ip"`
	Gateway string `json:"gateway"`
	DNS     string `json:"dns"`
}

type UploadMsg struct {
	Type string `json:"type"`
	Path string `json:"path"`
	Data string `json:"data"`
	Mode string `json:"mode"`
	Seq  int    `json:"seq"`
	Eof  bool   `json:"eof"`
}

type DownloadReqMsg struct {
	Type string `json:"type"`
	Path string `json:"path"`
}

type PtyExecRequest struct {
	Type    string            `json:"type"`
	ID      string            `json:"id"`
	Cmd     string            `json:"cmd"`
	Timeout int               `json:"timeout"`
	Env     map[string]string `json:"env"`
	Cols    int               `json:"cols"`
	Rows    int               `json:"rows"`
}

type PtyInputMsg struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Data string `json:"data"`
}

type WindowResizeMsg struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Cols int    `json:"cols"`
	Rows int    `json:"rows"`
}

type PortForwardStartMsg struct {
	Type      string `json:"type"`
	GuestPort int    `json:"guest_port"`
}

type PortForwardStopMsg struct {
	Type      string `json:"type"`
	GuestPort int    `json:"guest_port"`
}

// ── Guest → host messages ──────────────────────────────────────────

// ProtocolVersion is the wire-format version this agent speaks.
// Bump on any breaking wire change so the host can branch cleanly
// (or refuse to talk to a too-old/too-new agent) instead of
// guessing from message-shape probing. Old agents predate this
// field and the host treats absence as version=0.
const ProtocolVersion = 1

// AgentCapabilities is the conservative list of message families
// this build of the agent implements. The host may consult it
// before issuing a `pty_exec` or `port_forward_start` against an
// older guest. Adding a capability is non-breaking (older hosts
// ignore unknown strings); removing one IS breaking and must be
// paired with a ProtocolVersion bump.
var AgentCapabilities = []string{
	"pty",              // pty_exec / pty_input / pty_output / window_resize
	"port_forward",     // port_forward_start / stop / ready / error
	"network_metrics",  // periodic network_metrics frames
	"network_error",    // typed network_error on configure_network failure
	"binary_output",    // base64-encoded output for non-UTF-8 chunks
	"configure_network",
	"stdin",
}

type ReadyMsg struct {
	Type            string   `json:"type"`
	ProtocolVersion int      `json:"protocol_version,omitempty"`
	Capabilities    []string `json:"capabilities,omitempty"`
}

func NewReady() ReadyMsg {
	return ReadyMsg{
		Type:            TypeReady,
		ProtocolVersion: ProtocolVersion,
		Capabilities:    AgentCapabilities,
	}
}

type HeartbeatMsg struct {
	Type string `json:"type"`
}

func NewHeartbeat() HeartbeatMsg { return HeartbeatMsg{Type: TypeHeartbeat} }

type OutputMsg struct {
	Type     string `json:"type"`
	ID       string `json:"id"`
	Stream   string `json:"stream"`
	Data     string `json:"data"`
	Encoding string `json:"encoding,omitempty"` // "base64" for non-UTF-8 binary chunks; absent for text
}

type ExitMsg struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Code int    `json:"code"`
}

func NewExit(id string, code int) ExitMsg {
	return ExitMsg{Type: TypeExit, ID: id, Code: code}
}

type NetworkReadyMsg struct {
	Type string `json:"type"`
	IP   string `json:"ip"`
	// ConfigMs is the wall-clock time from configure_network receipt
	// to network_ready emission on the guest. Host uses this to
	// populate BootPhases.networkReadyMs without round-trip jitter
	// contaminating the measurement.
	ConfigMs int `json:"config_ms,omitempty"`
	// Stage annotates which readiness gate fired: "operstate",
	// "carrier", "route-verified", or "timeout-anyway". "timeout-
	// anyway" means the guest shipped network_ready on the
	// defensive fallback; host should treat this as a warning
	// signal, not a hard guarantee. Absent on pre-v0.7.1 agents.
	Stage string `json:"stage,omitempty"`
}

// InterfaceCounters is the per-NIC counter snapshot shipped inside
// NetworkMetricsMsg. Values are cumulative since interface-up, read
// straight from /proc/net/dev — the same numbers `ifconfig`/`ip -s`
// would show. Hosts compute deltas themselves if they want per-
// interval throughput.
type InterfaceCounters struct {
	RxBytes   uint64 `json:"rx_bytes"`
	TxBytes   uint64 `json:"tx_bytes"`
	RxErrors  uint64 `json:"rx_errors"`
	TxErrors  uint64 `json:"tx_errors"`
	RxPackets uint64 `json:"rx_packets"`
	TxPackets uint64 `json:"tx_packets"`
}

// NetworkMetricsMsg is a periodic snapshot of guest-side network
// counters. The map-shaped `interfaces` field is intentional: multi-
// NIC VMs are a foreseeable future, and renaming a singular `iface`
// field later would break the protocol. `lo` is excluded — consumers
// care about externally-visible traffic, not loopback.
type NetworkMetricsMsg struct {
	Type       string                       `json:"type"`
	Interfaces map[string]InterfaceCounters `json:"interfaces"`
}

// NetworkErrorMsg signals the guest failed to bring up the network.
// Sent when the `ip -batch` command failed AND the individual retry
// attempts also failed — the interface is unusable. Host propagates
// this to the waiting configureNetwork() caller as a typed error.
// Distinct from "timeout-anyway" which means network_ready fired on
// a softer guarantee; NetworkErrorMsg means the setup itself broke.
type NetworkErrorMsg struct {
	Type   string `json:"type"`
	Reason string `json:"reason"`
	// Attempts is how many retries the guest made before giving up.
	// Diagnostic only; host uses it for log prefix, not for retry
	// logic.
	Attempts int `json:"attempts,omitempty"`
}

type UploadAckMsg struct {
	Type string `json:"type"`
	Seq  int    `json:"seq"`
}

type UploadErrorMsg struct {
	Type  string `json:"type"`
	Path  string `json:"path"`
	Error string `json:"error"`
}

type UploadDoneMsg struct {
	Type string `json:"type"`
	Path string `json:"path"`
}

type DownloadDataMsg struct {
	Type string `json:"type"`
	Path string `json:"path"`
	Data string `json:"data"`
	Seq  int    `json:"seq"`
	Eof  bool   `json:"eof"`
}

type DownloadErrorMsg struct {
	Type  string `json:"type"`
	Path  string `json:"path"`
	Error string `json:"error"`
}

type PtyOutputMsg struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Data string `json:"data"`
}

type PortForwardReadyMsg struct {
	Type      string `json:"type"`
	GuestPort int    `json:"guest_port"`
	VsockPort int    `json:"vsock_port"`
}

// PortForwardErrorMsg signals the guest couldn't establish a forward.
// Sent in response to a PortForwardStartMsg that collided with an
// existing forward on the same guest port, or when the guest-side
// vsock bind failed. The host dispatches this to the pending
// portForwardContinuation and surfaces the reason to the caller.
//
// Reasons are human-readable strings — callers should not parse them,
// they are for logs and error messages. Well-known values today:
//   - "already active"    : host sent start twice for the same port
//   - "vsock listen"      : guest-side vsock bind failed
type PortForwardErrorMsg struct {
	Type      string `json:"type"`
	GuestPort int    `json:"guest_port"`
	Reason    string `json:"reason"`
}
